### ISSUE TITLE
Passing `:chdir` to `Kernel.system` doesn't work reliably

### ISSUE THREAD

**Issue by deivid-rodriguez** [jruby/jruby#6153]
**Environment Information**

```
$ jruby -v
jruby 9.2.11.0 (2.5.7) 2020-03-02 612d7a05a6 OpenJDK 64-Bit Server VM 11.0.6+10-post-Ubuntu-1ubuntu118.04.1 on 11.0.6+10-post-Ubuntu-1ubuntu118.04.1 [linux-x86_64]
$ env | grep JRUBY
JRUBY_OPTS= --dev --debug -J-Xmx1024m
$ env | grep JAVA
JAVA_HOME=/usr/lib/jvm/java-11-openjdk-amd64
$ uname -a
Linux chaba 5.3.0-42-generic #34~18.04.1-Ubuntu SMP Fri Feb 28 13:42:26 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

**Expected Behavior**

I expect that the following code would behave the same as common ruby, and print "hola" to the screen.

```
system(Gem.ruby, "-e", "puts('hola')", :chdir => Dir.pwd)
```

**Actual Behavior**

Instead, I get

```
sh: 1: Syntax error: "(" unexpected
```

Without the `:chdir` option, it works as expected.

**Comment by headius** [comment]
The root problem here is likely not the options but the `:chdir`.

Our native process launching is based on `posix_spawn`, which does not support doing `chdir` in the new child process. So this ends up falling back on JDK process launching, which does support `chdir`.

This means we have to hand-assemble a command line to pass to `sh -c` (Ruby launches via sh for these sorts of command lines), and that resulting command line probably does not properly escape the parentheses in `puts('hola')`.

The `chdir` behavior is not easy for us to fix; normally, if we could fork, make changes, and then exec, we'd chdir in the forked subprocess before calling exec (as CRuby and the JDK process APIs do). Unfortunately we can't safely fork from Java, so our options are no chdir and real native process launching, or chdir and Java's cumbersome process launching.

There are a couple workarounds for you:

* You can restructure your command line to be a valid sh script that does a chdir before starting.
* If you know that changing the directory in the parent process won't interfere with other executing threads, you can do a native chdir yourself before calling `system`.

**Comment by deivid-rodriguez** [comment]
Thank you!

I already workarounded it, but I thought it would be good to report it since it seemed like a clear bug.

**Comment by eregon** [comment]
FWIW in TruffleRuby we just use a small native executable (written in C) that `chdir()` and `execve()`.
I think that's a robust solution, it's also what the JDK did at some point.

**Comment by headius** [comment]
@eregon We do not ship any code that must be built on the target system.

**Comment by headius** [comment]
After chatting with @Mrnoname1000 a bit I think we can improve this.

He pointed out to me that you can invoke `sh` another way to avoid having to slam all the `system` arguments together into a single command line:

```
sh -c "chdir $0; exec $@" <command> <arg1> <arg2>
```

Or formatted like a call to `exec`:

```
"/bin/sh", "sh", "-c", "\"chdir $1; exec $@\"", "command", "arg1", "arg2"
```

I have not attempted to do this in our code yet but it may be a solution.

**Comment by headius** [comment]
@eregon Oh, FWIW... OpenJDK actually just uses fork+exec, via a single downcall into their C code. The bulk of the logic is in "forkAndExec".

**Comment by ahorek** [comment]
could we just use the native  ```runtime.getPosix().chdir(CharSequence path);``` which is already available in jruby?

https://github.com/oracle/truffleruby/blob/master/src/main/ruby/truffleruby/core/dir.rb#L301

**Comment by headius** [comment]
@ahorek Yes, subject to this caveat I mentioned above:

> * If you know that changing the directory in the parent process won't interfere with other executing threads, you can do a native chdir yourself before calling system.

**Comment by headius** [comment]
Ah to clarify something else... JRuby's `Dir.chdir` is *not* the native chdir, since it's not thread-safe and we don't know what other parts of the JVM may depend on the current directory of the process staying the same. JRuby simulates `chdir` by setting a per-runtime current directory and providing that to commands that would access files using `./` or process launches. When the JRuby runtime-level PWD doesn't match the JVM's we force a `chdir` similar to this case.

**Comment by headius** [comment]
Circling back to this again...

Here's examples from @Mrnoname1000 for how we might do this with `sh` and `chdir` without mucking up the other arguments:

https://gist.github.com/Mrnoname1000/bc40990c266793029e86919bb830c51a

The third option seems like the most appropriate since it doesn't depend on env and doesn't launch a second shell.

**Comment by headius** [comment]
My description is a bit confusing... the third version above avoids the intermediate "dangling" shell showing up as a process, since the subsequent exec of the new command takes over that process space.

**Comment by headius** [comment]
In exploring this I also realize an additional bug: `system` running without our native support and passing `chdir:` will ignore the chdir, because the old `system` logic doesn't know about it.

**Comment by headius** [comment]
Another nugget from @Mrnoname1000, this would be the safe way for us to stuff the `cd` call into an argument intended to be run as a shell script, a la something like `system("touch blah; do something; something else", chdir: "some dir")`:

chdir plus `-c` script without interpolating the dir (this is presented in the form of a C `execl` call):

```c
execl("/bin/sh", "/bin/sh", "-c" "cd \"$0\"; command_here", target_dir, NULL);
```

The primary technique is passing any command arguments as process-launch arguments rather than manually trying to stitch together a valid `sh` script.

**Comment by headius** [comment]
I have applied the first modification, for the original reported issue, in PR #6211. It works properly now and the embedded Ruby script is not interpreted as shell code.

These fixes will likely take some time to settle, since we are introducing process-launching patterns that were not originally designed into this code.

**Comment by headius** [comment]
Fixed by #6226.

**Comment by deivid-rodriguez** [comment]
Thanks for the fix!
